dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,488 @@
+            for h in SHA._history:
+                h["model_id"] = key(bracket, h["model_id"])
+                h["bracket"] = bracket
+        self.history_ = sum([SHA._history for SHA in SHAs.values()], [])
+
+        self.metadata_ = {
+            "models": sum(m["models"] for m in meta),
+            "partial_fit_calls": sum(m["partial_fit_calls"] for m in meta),
+            "brackets": meta,
+        }
+        raise gen.Return(self)
+
+    def metadata(self):
+        """Get information about how much computation is required for
+        :func:`~dask_ml.model_selection.HyperbandCV.fit`. This can be called
+        before or after  :func:`~dask_ml.model_selection.HyperbandCV.fit`.
+
+        Returns
+        -------
+        metadata : dict
+            Information about the computation performed by ``fit``. Has keys
+
+            * ``partial_fit_calls``, the total number of partial fit calls.
+            * ``models``, the total number of models created.
+            * ``brackets``, each of which has the same two keys as above.
+
+        Notes
+        ------
+        This algorithm runs several loops in an "embarassingly parallel"
+        manner. The ``brackets`` key represents each of these loops.
+
+        """
+        bracket_info = _hyperband_paper_alg(self.max_iter, eta=self.aggressiveness)
+        num_models = sum(b["models"] for b in bracket_info)
+        for bracket in bracket_info:
+            bracket["iters"].update({1})
+            bracket["iters"] = sorted(list(bracket["iters"]))
+        num_partial_fit = sum(b["partial_fit_calls"] for b in bracket_info)
+        bracket_info = reversed(sorted(bracket_info, key=lambda x: x["bracket"]))
+
+        info = {
+            "partial_fit_calls": num_partial_fit,
+            "models": num_models,
+            "brackets": list(bracket_info),
+        }
+        return info
+
+
+def _get_meta(hists, brackets, key=None):
+    if key is None:
+        key = lambda bracket, ident: "bracket={}-{}".format(bracket, ident)
+    meta_ = []
+    history_ = {}
+    for bracket in brackets:
+        hist = hists[bracket]
+
+        info_hist = {key(bracket, h["model_id"]): [] for h in hist}
+        for h in hist:
+            info_hist[key(bracket, h["model_id"])] += [dict(bracket=bracket, **h)]
+        hist = info_hist
+        history_.update(hist)
+
+        calls = {k: max(hi["partial_fit_calls"] for hi in h) for k, h in hist.items()}
+        iters = {hi["partial_fit_calls"] for h in hist.values() for hi in h}
+        meta_ += [
+            {
+                "bracket": "bracket=" + str(bracket),
+                "iters": sorted(list(iters)),
+                "models": len(hist),
+                "partial_fit_calls": sum(calls.values()),
+            }
+        ]
+    return meta_, history_
+
+
+def _get_cv_results(hists, params, key=None):
+    if key is None:
+        key = lambda bracket, ident: "bracket={}-{}".format(bracket, ident)
+    info = {
```

Comment:
@stsievert when you get a chance, could you explain this block of code (through line 415).

On line 405, we `zip(hist, params[bracket])`. These have different lengths


```python
In : [(len(x), len(params[bracket])) for (bracket, x) in hists.items()]
[(40, 27), (29, 12), (14, 6), (8, 4)]
```

There are tests that assert the number of fit calls is 49 (the number of items in the flattened `params.values()`).

Is this a bug, or do we know that the first elements of `hists[bracket]` match with the first elements of `params[bracket]`, so it's OK to drop the remained.

Reply:
For debugging, I've dropped a pdb on 417, and typed `interact` at the pdb console https://docs.python.org/3/library/pdb.html#pdbcommand-interact, which avoids some namespace issues with regular pdb). I'd never heard of `interact` before.

Reply:
> On line 405, we zip(hist, params[bracket]). These have different lengths

This is a bug. I don't see why the should be the same length: `hist` is a record of all calls to `additional_calls`, and `params` is only as long as the number of models.

The later part of this code (408-415) is the meat of this function: it records history for each model, in `{model_id: [params]}` format. I don't think 411 is necessary; I found it to be always true I think.

At 430, history is recorded per model: `info = {model_uid: [h1, h2, ...]}` where `h1` has keys specified by 396. From there, it's formatted to match sklearn's `cv_results_`.

Reply:
OK, thanks. I'll study that a bit more. I think my generic implementation at https://github.com/TomAugspurger/dask-ml/blob/stsievert-hyperband-2/dask_ml/model_selection/_incremental.py#L471

According to https://github.com/dask/dask-ml/pull/221/files#diff-40bbd5e21dc2b6e9bd0a31d9f21b81beR107 that should equal 49 (I think).